pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,435 @@
+	if scheduledEviction != nil {
+		startTime = scheduledEviction.CreatedAt
+		if startTime.Add(minTolerationTime).Before(triggerTime) {
+			return
+		} else {
+			tc.taintEvictionQueue.CancelWork(podNamespacedName.String())
+		}
+	}
+	tc.taintEvictionQueue.AddWork(NewWorkArgs(podNamespacedName.Name, podNamespacedName.Namespace), startTime, triggerTime)
+}
+
+func (tc *TaintController) handlePodUpdate(podUpdate podUpdateItem) {
+	// Create
+	if podUpdate.oldPod == nil {
+		pod := podUpdate.newPod
+		podNamespacedName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+		nodeName := pod.Spec.NodeName
+		glog.Infof("Noticed pod creation: %v", podNamespacedName.String())
+		if nodeName == "" {
+			return
+		}
+		taints, ok := tc.taintedNodes[nodeName]
+		if !ok {
+			return
+		}
+		tolerations := podUpdate.newTolerations
+		tolerated, matchingTolerations := getMatchingTolerations(taints, tolerations)
+		if tolerated {
+			return
+		}
+		forever := true
+		for i := range matchingTolerations {
+			if matchingTolerations[i].TolerationSeconds != nil {
+				forever = false
+				break
+			}
+		}
+		if !forever {
+			glog.Infof("Pod %v that doesn't tolerate NoExecute forever taint on Node %v was scheduled there.", podNamespacedName.String(), nodeName)
+			// It shouldn't have scheduled here in the first place - remove it right away
+			tc.taintEvictionQueue.AddWork(NewWorkArgs(podNamespacedName.Name, podNamespacedName.Namespace), time.Now(), time.Now())
+		}
+		return
+	}
+	// Delete
+	if podUpdate.newPod == nil {
+		pod := podUpdate.newPod
+		podNamespacedName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+		glog.Infof("Noticed pod deletion: %v", podNamespacedName.String())
+		tc.taintEvictionQueue.CancelWork(podNamespacedName.String())
+		return
+	}
+	// Update
+	pod := podUpdate.newPod
+	podNamespacedName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+	glog.Infof("Noticed pod update: %v", podNamespacedName.String())
+	nodeName := pod.Spec.NodeName
+	if nodeName == "" {
+		return
+	}
+	taints, ok := tc.taintedNodes[nodeName]
+	if !ok {
+		return
+	}
+	tolerations := podUpdate.newTolerations
+	tc.processPodOnNode(podNamespacedName, nodeName, tolerations, taints, time.Now())
+}
+
+func (tc *TaintController) handleNodeUpdate(nodeUpdate nodeUpdateItem) {
+	// Create
+	if nodeUpdate.oldNode == nil {
+		node := nodeUpdate.newNode
+		taints := nodeUpdate.newTaints
+		if len(taints) == 0 {
+			return
+		}
+		tc.taintedNodes[node.Name] = taints
+		pods := getPodsAssignedToNode(tc.client, node.Name)
+		if len(pods) == 0 {
+			return
+		}
+		now := time.Now()
+		for i := range pods {
+			tolerations, err := v1.GetPodTolerations(&pods[i])
+			if err != nil {
+				glog.Errorf("Failed to get tolerations from the Pod %v/%v, skipping", pods[i].Namespace, pods[i].Name)
+				continue
+			}
+			podNamespacedName := types.NamespacedName{Namespace: pods[i].Namespace, Name: pods[i].Name}
+			if len(tolerations) == 0 {
```

Comment:
Assume there is a node nodeA tainted with taint {foo=bar:NoSchedule}, and pod podA which has no tolerations is running on nodeA, according to the check here, it results in podA evicted.
Actually what we want is just evict pods those don't tolerate NoExecute taints of nodes, or am I missing something?